java/src/com/google/idea/blaze/java/run/hotswap/BlazeReloadFileAction.java

```diff
@@ -0,0 +1,232 @@
+                    File tempOutputDir;
+                    try {
+                        tempOutputDir = Files.createTempDirectory("IjBazelHotswap").toFile();
+                        LOGGER.debug("Temp dir for HotSwap artifacts is " + tempOutputDir.getPath());
+                        tempOutputDir.deleteOnExit();
+                    } catch (IOException e) {
+                        LOGGER.error("Failed creating temp directory for hotswap", e);
+                        return;
+                    }
+
+                    ProgressManager.getInstance().run(new Task.Backgroundable(project, "Preparing to hotswap", true) {
+                        @Override
+                        public void run(@NotNull ProgressIndicator progressIndicator) {
+                            // We want to be aware of entries like darwin_arm64-fastbuild/bin/some-api/lib-api.jar and not interested in darwin_arm64-fastbuild/external-api-0.0.654.jar
+                            // TODO: It is a good idea to skip non-output artifacts.
+                            Supplier<List<String>> artifactsForLogging = Suppliers.memoize(() -> outputs.artifacts.keySet().stream()
+                                    .filter(artifact -> artifact.indexOf("/") != artifact.lastIndexOf("/"))
+                                    .collect(Collectors.toList()));
+                            // With Gateway in some cases actions got lost, so we want to log this action for every invocation
+                            if (LOGGER.isDebugEnabled()) {
+                                LOGGER.debug(String.format("Searching for classes to HotSwap for class %s with package %s among outputs %s", vf.getPath(), jarDirectory, artifactsForLogging.get()));
+                            } else {
+                                LOGGER.info(String.format("Searching for classes to HotSwap for class %s with package %s", vf.getPath(), jarDirectory));
+                            }
+
+                            findAndCopyOutputFile(vf, tempOutputDir, jarDirectory, outputs);
+                            if (tempOutputDir.listFiles().length > 0) {
+                                LOGGER.info(String.format("Invoking HotSwap for entries %s with package %s", Arrays.toString(tempOutputDir.listFiles()), jarDirectory));
+                                hotswapFile(project, jarDirectory, tempOutputDir);
+                            } else {
+                                LOGGER.warn(String.format("There was nothing found for HotSwap for package %s in outputs %s", jarDirectory, artifactsForLogging.get()));
+                            }
+                        }
+                    });
+                } else {
+                    LOGGER.debug("Run hotswap after compile set to 'never'");
+                }
+
+            } catch (Exception e) {
+                LOGGER.error("Exception occurred during building file to hotswap", e);
```

Comment:
Action cancellation is not handled correctly. Cancelling an ongoing hotswap results in this internal IDE error being thrown:

```
Exception occurred during building file to hotswap

java.util.concurrent.CancellationException: Task was cancelled.
	at com.google.common.util.concurrent.AbstractFuture.cancellationExceptionWithCause(AbstractFuture.java:1543)
	at com.google.common.util.concurrent.AbstractFuture.getDoneValue(AbstractFuture.java:586)
	at com.google.common.util.concurrent.AbstractFuture.get(AbstractFuture.java:567)
	at com.google.common.util.concurrent.FluentFuture$TrustedFuture.get(FluentFuture.java:91)
	at com.google.idea.blaze.java.run.hotswap.BlazeReloadFileAction.lambda$actionPerformed$1(BlazeReloadFileAction.java:76)
	at com.intellij.openapi.application.impl.ApplicationImpl$2.run(ApplicationImpl.java:249)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at com.intellij.util.concurrency.ContextCallable.call(ContextCallable.java:32)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at com.intellij.util.concurrency.ContextRunnable.run(ContextRunnable.java:27)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:702)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:699)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1.run(Executors.java:699)
	at java.base/java.lang.Thread.run(Thread.java:840)
```